File: src/content/docs/topics/cs/coding-problems/linked-list/002-add-two-numbers.py
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_list(head):
    out = []
    while head:
        out.append(head.val)
        head = head.next
    return out

def from_list(vals):
    dummy = ListNode()
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next

def add_two_numbers(l1, l2):
    pass  # TODO: implement

def _run_tests():
    assert to_list(add_two_numbers(from_list([2, 4, 3]), from_list([5, 6, 4]))) == [7, 0, 8]
    assert to_list(add_two_numbers(from_list([0]), from_list([0]))) == [0]
    assert to_list(add_two_numbers(from_list([9, 9, 9, 9, 9, 9, 9]), from_list([9, 9, 9, 9]))) == [8, 9, 9, 9, 0, 0, 0, 1]
    assert to_list(add_two_numbers(from_list([1]), from_list([2]))) == [3]
    assert to_list(add_two_numbers(from_list([5]), from_list([5]))) == [0, 1]
    # --- large-input timing ---
    import time as _t
    def _make_list(vals):
        if not vals: return None
        head = ListNode(vals[0])
        cur = head
        for v in vals[1:]: cur.next = ListNode(v); cur = cur.next
        return head
    _head1 = _make_list(list(range(1000)))
    _head2 = _make_list(list(range(1000)))
    _t0 = _t.perf_counter()
    add_two_numbers(_head1, _head2)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf add_two_numbers(1000-digit numbers): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/linked-list/002-add-two-numbers.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val; this.next = next;
    }
}

function toList(head: ListNode | null): number[] {
    const out: number[] = [];
    while (head) { out.push(head.val); head = head.next; }
    return out;
}

function fromList(vals: number[]): ListNode | null {
    const dummy = new ListNode();
    let cur = dummy;
    for (const v of vals) { cur.next = new ListNode(v); cur = cur.next; }
    return dummy.next;
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    // TODO: implement
    return null;
}

function _runTests(): void {
    assert(JSON.stringify(toList(addTwoNumbers(fromList([2,4,3]), fromList([5,6,4])))) === JSON.stringify([7,0,8]));
    assert(JSON.stringify(toList(addTwoNumbers(fromList([0]), fromList([0])))) === JSON.stringify([0]));
    assert(JSON.stringify(toList(addTwoNumbers(fromList([9,9,9,9,9,9,9]), fromList([9,9,9,9])))) === JSON.stringify([8,9,9,9,0,0,0,1]));
    assert(JSON.stringify(toList(addTwoNumbers(fromList([1]), fromList([2])))) === JSON.stringify([3]));
    assert(JSON.stringify(toList(addTwoNumbers(fromList([5]), fromList([5])))) === JSON.stringify([0,1]));
    // perf
    const head1 = fromList(Array.from({ length: 100_000 }, () => 9));
    const head2 = fromList(Array.from({ length: 100_000 }, () => 9));
    const t0 = performance.now();
    addTwoNumbers(head1, head2);
    console.log(`perf addTwoNumbers(100000-digit numbers): ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
